mikupad.html

```diff
@@ -2014,6 +2070,161 @@
+        y: dragRef.current.initialY + deltaY,
+      });
+    };
+
+    const handleMouseUp = () => {
+      setIsDragging(false);
+    };
+
+    document.addEventListener('mousemove', handleMouseMove);
+    document.addEventListener('mouseup', handleMouseUp);
+
+    return () => {
+      document.removeEventListener('mousemove', handleMouseMove);
+      document.removeEventListener('mouseup', handleMouseUp);
+    };
+  }, [isDragging]);
+
+  const handleMouseDown = (e) => {
+    setIsDragging(true);
+    dragRef.current = {
+      startX: e.clientX,
+      startY: e.clientY,
+      initialX: position.x,
+      initialY: position.y,
+    };
+  };
+
+	useEffect(() => {
+		const onKeyDown = (event) => {
+			if (event.key === 'Escape') {
+				onClose();
+			}
+		};
+		document.addEventListener('keydown', onKeyDown);
+		return () => {
+			document.removeEventListener('keydown', onKeyDown);
+		};
+	}, []);
+
+	return html`
+		<div className="widget-body"
+			id="${id}"
+			style=${{ transform: `translate(${position.x}px, ${position.y}px)` }}>
+			<div class="widget-container">
+				<div class="widget-title"
+					style=${{ cursor: isDragging ? 'grabbing' : 'grab' }}
+					onMouseDown=${handleMouseDown}
+					onMouseMove=${() => {}}
+					onMouseUp=${() => {}}
+					onMouseLeave=${() => {}}>
+					${title}
+				</div>
+				<div className="widget-content">
+					${children}
+				</div>
+				<button
+					class="button-widget-top"
+					onClick=${onClose}>
+					<svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="-1 -1 10 10" xmlns="http://www.w3.org/2000/svg"><path d="M 0 1 L 3 4 L 0 7 L 1 8 L 4 5 L 7 8 L 8 7 L 5 4 L 8 1 L 7 0 L 4 3 L 1 0 L 1 0 Z"></path></svg>
+				</button>
+			</div>
+		</div>`;
+}
+function SearchAndReplaceWidget({ isOpen, closeWidget, id, children, promptArea, cancel, ...props }) {
+	const [searchAndReplaceError, setSearchAndReplaceError] = useState(undefined);
+	const [searchAndReplaceMode, setSearchAndReplaceMode] = usePersistentState('searchAndReplaceMode', 0);
+	const [searchTerm, setSearchTerm] = usePersistentState('searchTerm', '');
+	const [replaceTerm, setReplaceTerm] = usePersistentState('replaceTerm', '');
+	const [inputElement, setInputElement] = useState(null);
+
+	useEffect(() => {
+		if (promptArea.current) {
+			setInputElement(promptArea.current);
+		}
+	}, [promptArea]);
+
+	function handleSearchAndReplace(mode,search,replace) {
+		setSearchAndReplaceError(undefined)
+		// console.log(prompt)
+		if (!search)
+			return
+
+		switch(mode) {
+			case 0:
+				plaintextReplace(search,replace,inputElement)
+				break;
+			case 1:
+				regexReplace(search,replace,inputElement)
+				break;
+			case 2:
+				templateReplace(search,replace,inputElement)
+				break;
+		}
+	}
+
+	function plaintextReplace(search,replace,elem) {
+		// console.log(elem.value)
+		// need to figure out a smart way to keep the cursor position
+		elem.value = elem.value.replaceAll(search,replace)

```

Comment:
And I could actually already use some help. In`SearchAndReplaceWidget` I tried to import promptArea, but I can't seem to use it in my functions unless I do weird stuff with useEffect a little further up.
Problem is that now, after replacing, the changes revert on predicting or reloading unless you do a manual edit first, and I don't have the slightest clue how to fix this.

Reply:
I'm not sure why you can't import promptArea, but the issue with the changes reverting is simple to solve: You need to trigger the input event. You would have to either somehow call onInput like [this](https://github.com/lmg-anon/mikupad/blob/d5ba8962af85b618d4948ee42da7996eeec15d2c/mikupad.html#L3952), or trigger the input event like this:

```js
const event = new Event('input', { bubbles: true });
promptArea.current.dispatchEvent(event);
```"

Reply:
Right, that works. Thanks!